# Student Performance Indicator

### LIfe Cycle of a Machine Learning Project

* Understanding the Problem Statement
* Data Collection 
* Data Validation
* Exploratory Data Analysis 
* Data Pre-Processing
* Model Training
* Choose Best Model

## 1.) Problem Statement

* This project understands how the student's performance (test scores) is affected by the other variables such as Gender, Ethnicity, Parental level of education, Lunch, and Test preparation course.

## 2.) Data Collection

* Data Source - https://www.kaggle.com/datasets/spscientist/students-performance-in-exams?datasetId=74977
* The data consists of * columns and 1000 rows. 

## 2.1) Import Data and Required Packages

In [23]:
#Import Dependencies 
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline
import warnings
warnings.filterwarnings('ignore')

## Extract CSV Data into Pandas Dataframe

In [24]:
#Create Student Performance Dataframe
sp_df = pd.read_csv('data/stud.csv')

#View Dataframe 
sp_df.head()

gender race_ethnicity parental_level_of_education         lunch  \
0  female        group B           bachelor's degree      standard   
1  female        group C                some college      standard   
2  female        group B             master's degree      standard   
3    male        group A          associate's degree  free/reduced   
4    male        group C                some college      standard   

  test_preparation_course  math_score  reading_score  writing_score  
0                    none          72             72             74  
1               completed          69             90             88  
2                    none          90             95             93  
3                    none          47             57             44  
4                    none          76             78             75

In [25]:
#View Dataframe Shape
sp_df.shape

(1000, 8)

## 2.2) Dataset Information 

* gender: Sex of Student -> (male/female)
* race_ethnicity: Ethnicity of the Student -> (Group A, B, C, D, E)
* parental_level_of_education: Parent's highest education completed -> (master's degree, bachelor's degree, associate's degree, some college, high school)
* lunch: Type of lunch -> (standard, free/reduced)
* test_preparation_course: supplemental material completion -> (completed, none)
* math_score
* reading_score
* writing_score  

## 3.) Data Validation

* Check missing values
* Check duplicates
* Check data types
* Check the number of unique values in each column
* Check statistics of the dataset
* Check various categories present in the columns 

## 3.1.) Check for Missing Values

In [26]:
#Count the amount of missing values
sp_df.isna().sum()

gender                         0
race_ethnicity                 0
parental_level_of_education    0
lunch                          0
test_preparation_course        0
math_score                     0
reading_score                  0
writing_score                  0
dtype: int64

* There are no missing values in the dataset

## 3.2.) Check for Duplicate Values

In [27]:
#Count the amount of duplicate values
sp_df.duplicated().sum()

0

* There are no duplicate values in the dataset

## 3.3.) Check Data Types

In [28]:
#Check for null values, and the data types in the dataset
sp_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 8 columns):
 #   Column                       Non-Null Count  Dtype 
---  ------                       --------------  ----- 
 0   gender                       1000 non-null   object
 1   race_ethnicity               1000 non-null   object
 2   parental_level_of_education  1000 non-null   object
 3   lunch                        1000 non-null   object
 4   test_preparation_course      1000 non-null   object
 5   math_score                   1000 non-null   int64 
 6   reading_score                1000 non-null   int64 
 7   writing_score                1000 non-null   int64 
dtypes: int64(3), object(5)
memory usage: 62.6+ KB


* There are no null values in the dataset

## 3.4.) Check the number of Unique Values in the Dataset's columns

In [29]:
#Check dataset for unique values
sp_df.nunique()

gender                          2
race_ethnicity                  5
parental_level_of_education     6
lunch                           2
test_preparation_course         2
math_score                     81
reading_score                  72
writing_score                  77
dtype: int64

## 3.5.) Check the Statistics of the Dataset

In [30]:
#Check basic stats of the dataset
sp_df.describe()

math_score  reading_score  writing_score
count  1000.00000    1000.000000    1000.000000
mean     66.08900      69.169000      68.054000
std      15.16308      14.600192      15.195657
min       0.00000      17.000000      10.000000
25%      57.00000      59.000000      57.750000
50%      66.00000      70.000000      69.000000
75%      77.00000      79.000000      79.000000
max     100.00000     100.000000     100.000000

## Insights

* From the description above of the numerical data, all means are very similar - between 65 and 68.05;
* All standard deviations are also very similar - between 14.6 and 15.19;
* While there is a minimum score of 0 for the math portion of the exam, the minimum socre for the writing portion is 10, and 17 for the reading portion of the exam

## 3.6) Exploring the Data

In [31]:
#View gender categories in the dataset
print("Categories in 'gender' variables:    ", end = " ")
print(sp_df['gender'].unique())

#View race categories in the dataset
print("Categories in 'race_ethnicity' variables:    ", end = " ")
print(sp_df['race_ethnicity'].unique())

#View parental education categories in the dataset
print("Categories in 'parental_level_of_education' variables:    ", end = " ")
print(sp_df['parental_level_of_education'].unique())

#View lunch categories in the dataset
print("Categories in 'lunch' variables:    ", end = " ")
print(sp_df['lunch'].unique())

#View test prep categories in the dataset
print("Categories in 'test_preparation_course' variables:    ", end = " ")
print(sp_df['test_preparation_course'].unique())

Categories in 'gender' variables:     ['female' 'male']
Categories in 'race_ethnicity' variables:     ['group B' 'group C' 'group A' 'group D' 'group E']
Categories in 'parental_level_of_education' variables:     ["bachelor's degree" 'some college' "master's degree" "associate's degree"
 'high school' 'some high school']
Categories in 'lunch' variables:     ['standard' 'free/reduced']
Categories in 'test_preparation_course' variables:     ['none' 'completed']


In [32]:
#Define numerical and Categorical columns in dataset
numeric_features = [feature for feature in sp_df.columns if sp_df[feature].dtype != 'O']
categorical_features = [feature for feature in sp_df.columns if sp_df[feature].dtype == 'O']

#View numerical and categorical columns in dataset
print("There are {} numerical features: {}".format(len(numeric_features), numeric_features))
print("There are {} categorical features: {}".format(len(categorical_features), categorical_features))

There are 3 numerical features: ['math_score', 'reading_score', 'writing_score']
There are 5 categorical features: ['gender', 'race_ethnicity', 'parental_level_of_education', 'lunch', 'test_preparation_course']


## 3.7.) Creating Total score and Average Columns

In [33]:
#Create total score column
sp_df['total_score'] = sp_df['math_score'] + sp_df['reading_score'] + sp_df['writing_score']

#Create average column
sp_df['average'] = sp_df['total_score'] / 3

#View updated dataframe 
sp_df.head()

gender race_ethnicity parental_level_of_education         lunch  \
0  female        group B           bachelor's degree      standard   
1  female        group C                some college      standard   
2  female        group B             master's degree      standard   
3    male        group A          associate's degree  free/reduced   
4    male        group C                some college      standard   

  test_preparation_course  math_score  reading_score  writing_score  \
0                    none          72             72             74   
1               completed          69             90             88   
2                    none          90             95             93   
3                    none          47             57             44   
4                    none          76             78             75   

   total_score    average  
0          218  72.666667  
1          247  82.333333  
2          278  92.666667  
3          148  49.333333  
4          229  76.333333

## 3.8.) Explore test statistics

In [34]:
#Check the amount of students who recieved perfect test scores
math_perf = sp_df[sp_df['math_score'] == 100]['average'].count()
reading_perf = sp_df[sp_df['reading_score'] == 100]['average'].count()
writing_perf = sp_df[sp_df['writing_score'] == 100]['average'].count()

#View the amount of Students who recieved perfect test scores
print(f'Number of Students with perfect scores in math: {math_perf}')
print(f'Number of Students with perfect scores in reading: {reading_perf}')
print(f'Number of Students with perfect scores in writing: {writing_perf}')

Number of Students with perfect scores in math: 7
Number of Students with perfect scores in reading: 17
Number of Students with perfect scores in writing: 14


In [36]:
#Check the amount of students who recieved 20 or less on test scores
math_20 = sp_df[sp_df['math_score'] <= 20]['average'].count()
reading_20 = sp_df[sp_df['reading_score'] <= 20]['average'].count()
writing_20 = sp_df[sp_df['writing_score'] <= 20]['average'].count()

#View the amount of students who recieved 20 or less on test scores
print(f'Number of Students with math scores of 20 or less: {math_20}')
print(f'Numner of students with reading scores of 20 or less: {reading_20}')
print(f'Number of Students with writing scores of 20 or less: {writing_20}')

Number of Students with math scores of 20 or less: 4
Numner of students with reading scores of 20 or less: 1
Number of Students with writing scores of 20 or less: 3


## Insights

* From the values above we can conclude that students performed worse in math than in reading, and writing;
* We can also conclude that the best performance was seen in the reading portion of the exam

## 4.) Data Visualization 

## 4.1.) Visualize Average Score Distribution to Draw Conclusions 

* Create a Histogram 
* Create Kernel Distribution Function (KDE)

In [ ]:
#Create histogram visualization
